sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -116,6 +116,78 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     }
   }
 
+  override def visitBatchOrSingleStatement(ctx: BatchOrSingleStatementContext): BatchBody = {
+    if (ctx.batchCompound() != null) {
+      visit(ctx.batchCompound()).asInstanceOf[BatchBody]
+    } else {
+      val logicalPlan = visitSingleStatement(ctx.singleStatement())
+      BatchBody(List(SparkStatementWithPlan(
+        parsedPlan = logicalPlan,
+        sourceStart = ctx.start.getStartIndex,
+        sourceEnd = ctx.stop.getStopIndex + 1)))
+    }
+  }
+
+  override def visitBatchCompound(ctx: BatchCompoundContext): BatchBody = {
+    visitBatchBody(ctx.batchBody(), allowDeclareAtTop = true)
+  }
+
+  private def visitBatchBody(ctx: BatchBodyContext, allowDeclareAtTop: Boolean): BatchBody = {
+    val buff = ListBuffer[BatchPlanStatement]()
+    for (i <- 0 until ctx.getChildCount) {
+      val child = visit(ctx.getChild(i))
+      child match {
+        case statement: BatchPlanStatement => buff += statement
+        case null => // When terminal nodes are visited (like SEMICOLON, EOF, etc.)
```

Comment:
it would be good if this would be some cleaner check than generic null.

Reply:
Improved - didn't find a better way to do it, but I think it's clean like this.
Any other way I think we would need to match child before visiting and than again after visiting which would turn out to be really messy.